\begin{abstract}

In high-diversity populations, a complete accounting of \textit{de novo} mutations can be difficult to obtain.  Most analyses involve alignment of genomic reads to a reference genome, but if the haplotypic background upon which a mutation occurs is absent, events can be easily missed (as reads have nowhere to align) and false-positives may abound (as the aligner forces the reads to align elsewhere).  In this thesis, I describe methods for \textit{de novo} mutation discovery and genotyping based on a so-called "pedigree graph" - a de Bruijn graph where all available sequencing data (trusted and untrusted data alike) is represented.  I constrain genotyping efforts to locations containing "novel kmers" - sequence present in the child but absent from the parents.  I then apply Dijkstra's shortest path algorithm to perform the genotyping, even in the presence of sequencing error.  In simulation, this approach provides a vastly more sensitive and specific set of \textit{de novo} variants than traditional methods.

\end{abstract}
