\begin{abstract}

In high-diversity populations, a complete accounting of \textit{de novo} mutations can be difficult to obtain.  Most analyses involve alignment of genomic reads to a reference genome, but if the haplotypic background upon which a mutation occurs is absent, events can be easily missed (as reads have nowhere to align) and false-positives may abound (as the aligner forces the reads to align elsewhere).  In this thesis, we describe methods for \textit{de novo} mutation discovery and genotyping based on a multi-color de Bruijn graph where all available sequencing data (trusted and untrusted data alike) for members of an experimental cross or pedigree is represented.  We constrain variant discovery efforts to locations containing "novel kmers" - sequence present in the child but absent from the parents.  We then apply Dijkstra's shortest path algorithm to perform the genotyping, even in the presence of sequencing error.  Both simulation and validation data show this approach provides a vastly more sensitive and specific set of \textit{de novo} variants than traditional methods.  We apply our caller to real datasets: experimental crosses of \textit{Plasmodium falciparum} (the causal agent of malaria, $23$ Mbp), and members of a \textit{Pan troglodytes} pedigree (chimpanzee, $3,300$ Mbp genome).  Our approach is fast, flexible, accurate, and provides insight into regions of the genome completely inaccessible by reference-based methods.

\end{abstract}
